player/command: add udata property

rcombs opened the pull request:
This will replace shared-script-properties in new usage. It can be used for all the same things, but is much more versatile. Clients can create arbitrary sub-objects, and text expansion can access them.

For instance, if a script sets `udata/my-script/property1` to "test", that value will be available by expanding `${udata/my-script/property1}`.

sfan5 commented on player/command.c:
```diff
@@ -3558,6 +3563,103 @@ static int mp_property_script_props(void *ctx, struct m_property *prop,
+    struct m_property *prop = ctx->prop;
+    mpv_node *node = ctx->node;
+
+    if (node && node->format == MPV_FORMAT_NODE_ARRAY &&
+        (action == M_PROPERTY_KEY_ACTION || action == M_PROPERTY_GET ||
+         action == M_PROPERTY_GET_NODE))
+        return m_property_read_list(action, arg, node->u.list->num, &do_list_user_data, ctx);
```
A few more comments around here would be nice too.

rcombs replied:
Added.

sfan5 commented:
Also missing docs.

rcombs commented on misc/json.c:
```diff
@@ -300,12 +300,15 @@ static int json_append(bstr *b, const struct mpv_node *src, int indent)
+        if (indent == 0)
+            APPEND(b, src->u.string);
+        else
+            write_json_str(b, src->u.string);
```
This is only used for stringizing properties, and `indent == 0` means this only happens when (a) we're in "print" mode (i.e. pretty-printing, non-raw-mode text expansion), and (b) this is the top-level object. So basically, it means that directly pretty-printing the value of a string node will just give the string content, rather than the escaped/quote-wrapped version. This is substantially more useful for parameter expansion, and still allows using raw mode if you need something guaranteed to be parseable as JSON.

sfan5 replied:
If this is only used during pretty printing then that's fine, but I'm pretty sure this code is also used for IPC so it needs to be valid json.

rcombs replied:
`json_write` is used elsewhere (including in IPC), but `json_write_pretty` is used only via `m_option_pretty_print`, which is for human viewing, and the `indent == 0` case only occurs in `json_write_pretty`.

rcombs commented on player/command.c:
```diff
@@ -3558,6 +3563,103 @@ static int mp_property_script_props(void *ctx, struct m_property *prop,
+static int mp_property_user_data(void *ctx, struct m_property *prop,
+                                  int action, void *arg)
+{
+    if (action == M_PROPERTY_SET || action == M_PROPERTY_SET_NODE)
```
Done.

rcombs commented on misc/node.c:
```diff
@@ -45,12 +45,18 @@ struct mpv_node *node_array_add(struct mpv_node *dst, int format)
     MP_TARRAY_GROW(list, list->values, list->num);
     MP_TARRAY_GROW(list, list->keys, list->num);
-    list->keys[list->num] = talloc_strdup(list, key);
+    list->keys[list->num] = talloc_strndup(list, key.start, key.len);
```
I used `bstrdup0` instead, since `str.start` can't be null here.

rcombs commented:
Re-pushed with comments addressed, and additionally:
- Added the ability to observe sub-properties efficiently
- Added Lua and JS helpers
- Added docs

rcombs commented:
Re-pushed with:
- Renamed the new property from `user-data` to `udata`
- Added `del` command and various convenience functions to call it
- Implemented property deletion

sfan5 commented on player/client.c:
```diff
@@ -1586,6 +1586,37 @@ int mpv_unobserve_property(mpv_handle *ctx, uint64_t userdata)
+        bstr tmp = a;
+        a = b;
+        b = tmp;
+    }
```
might be worth shortcutting the common case here where `a.len == b.len` (that's common, right?)

rcombs replied:
Not sure what you mean?

sfan5 replied:
```c
} else if (a.len == b.len) {
    return bstrcmp(a, b) == 0;
}
```

rcombs replied:
I've rewritten this function to be a bit simpler.

sfan5 commented on options/m_property.h:
```diff
@@ -86,6 +86,11 @@ enum mp_property_action {
+    // Delete a value.
+    // Most properties do not implement this.
+    //  arg: (ignored)
+    M_PROPERTY_DELETE,
```
for the sake of udata, couldn't `M_PROPERTY_SET + { .format = MPV_FORMAT_NONE }` be made to do the same?

rcombs replied:
No, that's distinct; it sets a `MPV_FORMAT_NONE` value (a JSON `null` when serialized), which could be useful to distinguish from not-present in some situations.

sfan5 commented on player/command.c:
```diff
@@ -3760,6 +3927,7 @@ static const struct m_property mp_properties_base[] = {
     {"input-bindings", mp_property_bindings},
 
     {"shared-script-properties", mp_property_script_props},
+    {"udata", mp_property_udata},
```
my bikeshed contribution: either `userdata` or `user` but abbrevations are meh

garoto replied:
> my bikeshed contribution: either `userdata` or `user` but abbrevations are meh

`udata` is just plain bad, not a bikeshed at all.

Dudemanguy replied:
Just recording my bikeshed here, but I prefer `userdata`. It's very clear to me what that does as opposed to `udata` which makes me think of μ like utorrent or something.
